### PR TITLE
Update PCG method to support `LinearSolver` and callable functions as input argument

### PR DESCRIPTION
**Pull request by davschneller**
This PR slightly modifies the `psydac.linalg.iterative_solvers.pcg`, so that we may use a `LinearSolver` object or a callable function instead of a string, or simply input `None` as preconditioner. (note that as of now, this PR modifies as little as necessary)

* If we input no preconditioner (i.e. `pc is None`), the regular `cg` method is called.
* If a `LinearSolver` is used, it utilizes the `out` parameter of the `solve` method, i.e. we do not allocate a new vector for each round.
* If a callable function is used, the parameter format is as for the functions which are addressable as strings. I.e. we have `function(A, r)`, where we want to solve `As = r`. `s` is always re-allocated.
* The two string parameters `'jacobi', 'weighted_jacobi'` are kept for compatibility reasons with the existing code for now.
* Includes an interface test.

### CODE REVIEW

**Comment by davschneller** [comment]
I also made a branch `devel-cginterface` which changes a bit more of the code, i.e. it merges the `cg` and `pcg` methods, and adds more type checks (and also, it adds an out parameter). If wanted, I can add the changes done there to this PR.

File: psydac/linalg/iterative_solvers.py
Comment (yguclu) [inline psydac/linalg/iterative_solvers.py:178]: Please update the function's docstring after these changes.

**Review by yguclu — CHANGES_REQUESTED** [review]
This PR looks very good to me, and it can be merged once you update the docstring of function `pcg`.

**Comment by yguclu** [comment]
> I also made a branch `devel-cginterface` which changes a bit more of the code, i.e. it merges the `cg` and `pcg` methods, and adds more type checks (and also, it adds an out parameter). If wanted, I can add the changes done there to this PR.

@davschneller If you want to create an additional PR for branch `devel-cginterface`, I will be happy to have a look at it after this PR is merged

**Comment by davschneller** [comment]
Yes, thanks; I'll see if I'll make it

**Review by yguclu — APPROVED** [review]
I like these changes very much! in particular the fact that you avoid creating a temporary array at each call of the preconditioner.

**Pull request merged**
